server/api/src/rsDataObjRead.cpp

```diff
@@ -55,40 +61,70 @@ applyRuleForPostProcForRead( rsComm_t *rsComm, bytesBuf_t *dataObjReadOutBBuf, c
 }
 
 int
-rsDataObjRead( rsComm_t *rsComm, openedDataObjInp_t *dataObjReadInp,
-               bytesBuf_t *dataObjReadOutBBuf ) {
-    int bytesRead;
-    int l1descInx = dataObjReadInp->l1descInx;
-
-    if ( l1descInx < 2 || l1descInx >= NUM_L1_DESC ) {
-        rodsLog( LOG_NOTICE,
-                 "rsDataObjRead: l1descInx %d out of range",
-                 l1descInx );
+rsDataObjRead(rsComm_t* rsComm,
+              openedDataObjInp_t* dataObjReadInp,
+              bytesBuf_t* dataObjReadOutBBuf)
+{
+    const int l1descInx = dataObjReadInp->l1descInx;
+
+    if (l1descInx <= 2 || l1descInx >= NUM_L1_DESC) {
+        rodsLog(LOG_ERROR, "%s: l1descInx %d out of range", __func__, l1descInx);
         return SYS_FILE_DESC_OUT_OF_RANGE;
     }
-    if ( L1desc[l1descInx].inuseFlag != FD_INUSE ) {
+
+    auto& l1desc = L1desc[l1descInx];
+
+    if (l1desc.inuseFlag != FD_INUSE) {
         return BAD_INPUT_DESC_INDEX;
     }
 
-    if ( L1desc[l1descInx].remoteZoneHost != NULL ) {
-        /* cross zone operation */
-        dataObjReadInp->l1descInx = L1desc[l1descInx].remoteL1descInx;
-        bytesRead = rcDataObjRead( L1desc[l1descInx].remoteZoneHost->conn,
-                                   dataObjReadInp, dataObjReadOutBBuf );
+    if (l1desc.remoteZoneHost) {
+        // Cross zone operation.
+        dataObjReadInp->l1descInx = l1desc.remoteL1descInx;
+        const auto bytes_read = rcDataObjRead(l1desc.remoteZoneHost->conn, dataObjReadInp, dataObjReadOutBBuf);
         dataObjReadInp->l1descInx = l1descInx;
+        return bytes_read;
     }
-    else {
-        int i;
-        bytesRead = l3Read( rsComm, l1descInx, dataObjReadInp->len,
-                            dataObjReadOutBBuf );
-        i = applyRuleForPostProcForRead( rsComm, dataObjReadOutBBuf,
-                                         L1desc[l1descInx].dataObjInfo->objPath );
-        if ( i < 0 ) {
-            return i;
+
+    auto* dataObjInfo = l1desc.dataObjInfo;
+
+    // If the replica was opened in read-only mode, then don't allow the client
+    // to read past the data size in the catalog. This is necessary because the
+    // size of the replica in storage could be larger than the data size in the
+    // catalog.
+    //
+    // For all other modes, let the read operation do what it normally does.
+    if (!dataObjInfo->specColl && O_RDONLY == (l1desc.dataObjInp->openFlags & O_ACCMODE)) {
```

Comment:
Ignore this check if the special collections pointer is valid. This change was due to failing CI tests.

Does this seem reasonable?

Reply:
Also, is using `getStructFileType(dataObjInfo->specColl) >= 0` better than just checking the pointer for special collections?

What is the difference?

Reply:
The struct file check is kind of a subset of special collections. It's a type of special collection which represents a mounted struct file (e.g. tar). I've struggled with the difference myself because it seems that sometimes the struct files do things like regular data objects but at other times they don't, so I'm not 100% sure when it is appropriate to use one over the other.

TL;DR: if you just want to avoid special collections here, checking the `specColl` pointer should be sufficient

Reply:
Ok. Marking as resolved.